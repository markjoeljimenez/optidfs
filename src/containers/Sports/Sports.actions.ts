export const UPDATE_SPORT = 'UPDATE_SPORT';
export const SET_SPORTS = 'SET_SPORTS';
export const RESET_SELECTED_SPORT = 'RESET_SELECTED_SPORT';
