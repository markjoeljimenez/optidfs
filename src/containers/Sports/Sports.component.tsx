import { useDispatch } from 'react-redux';
import Select, { IValueLabel } from '../../components/form/select';
import { useAppSelector } from '../../hooks';

import { DROPDOWN_ACTIONS } from '../Dropdown/Dropdown.actions';

const Sports = () => {
	const { sports } = useAppSelector((state) => state);
	const dispatch = useDispatch();

	function handleSportChange(e: React.ChangeEvent<HTMLSelectElement>) {
		const sportId = parseInt(e.currentTarget.value);

		dispatch({
			type: DROPDOWN_ACTIONS.FETCH_CONTESTS,
			sport: sports.allSports.find(
				(_sport) => _sport.sportId === sportId
			),
		});
	}

	return (
		<div className="relative">
			<Select
				options={sports.allSports
					?.filter(
						(sport) =>
							sport.isEnabled &&
							sport.hasPublicContests &&
							sport.supported
					)
					.sort((a, b) =>
						a.regionAbbreviatedSportName.localeCompare(
							b.regionAbbreviatedSportName
						)
					)
					.map(
						(sport) =>
							({
								value: sport.sportId,
								label: sport.regionAbbreviatedSportName,
							} as IValueLabel)
					)}
				defaultValue=""
				hideLabel
				id="selectSport"
				label="Select sport"
				placeholder="Select sport"
				onChange={handleSportChange}
			/>
		</div>
	);
};

export default Sports;
