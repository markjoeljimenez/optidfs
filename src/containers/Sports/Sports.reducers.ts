import { HYDRATE } from 'next-redux-wrapper';
import { ISport } from '../../interfaces/ISports';
import {
	RESET_SELECTED_SPORT,
	SET_SPORTS,
	UPDATE_SPORT,
} from './Sports.actions';

interface ISportsState {
	allSports: ISport[];
	selectedSport?: ISport;
}

interface SportsAction extends ISportsState {
	type: string;
	payload: any;
}

const DEFAULT_STATE: ISportsState = {
	allSports: [],
};

const SportsReducer = (
	state = DEFAULT_STATE,
	{ type, allSports, selectedSport, payload }: SportsAction
) => {
	switch (type) {
		case HYDRATE:
			return {
				...state,
				allSports: payload.sports.allSports as ISport[],
			};

		case UPDATE_SPORT:
			return {
				...state,
				selectedSport,
			};

		case SET_SPORTS:
			return {
				...state,
				allSports,
			};

		case RESET_SELECTED_SPORT:
			return DEFAULT_STATE;

		default:
			return state;
	}
};

export default SportsReducer;
