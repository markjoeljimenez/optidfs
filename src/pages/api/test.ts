import { XMLParser } from 'fast-xml-parser';
import { NextApiRequest, NextApiResponse } from 'next';
import { getToken } from 'next-auth/jwt';

export default async function GET(req: NextApiRequest, res: NextApiResponse) {
	const token = await getToken({ req });

	const response = await fetch(
		`https://fantasysports.yahooapis.com/fantasy/v2/users;use_login=1/games`,
		{
			headers: { Authorization: `Bearer ${token?.accessToken}` },
		}
	);

	res.status(200).json(new XMLParser().parse(await response.text()));
}
