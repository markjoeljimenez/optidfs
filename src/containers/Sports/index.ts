import { getSportsFromProvider } from './api/getSportsFromProvider';
import sportsHandler from './handlers/sports.handler';
import sportsMock from './mocks/sports.mocks';
import SportsReducers, {
	sportsState,
	setSelectedSport,
} from './redux/reducers';
import Sports from './components/Sports';

// Api
export { getSportsFromProvider };

// Handlers
export { sportsHandler };

// Mocks
export { sportsMock };

// Reducers
export { SportsReducers };

// Actions
export { setSelectedSport };

// State
export { sportsState };

// Components
export default Sports;
