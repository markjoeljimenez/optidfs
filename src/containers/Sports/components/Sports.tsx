import { useAppDispatch, useAppSelector } from '../../../hooks';
import Select, { IValueLabel } from '../../../components/form/select';
import { sportsState, setSelectedSport } from '../redux/reducers';
import { useGetSportsFromProviderQuery } from '../../../api';
import { skipToken } from '@reduxjs/toolkit/dist/query';
import { selectProviders } from '@/containers/Providers';

const Sports = () => {
	const dispatch = useAppDispatch();
	const providers = useAppSelector(selectProviders);
	const sports = useAppSelector(sportsState);

	const test = useGetSportsFromProviderQuery(providers.provider ?? skipToken);

	// console.log(providers.provider, test);

	function handleSportChange(e: React.ChangeEvent<HTMLSelectElement>) {
		const sportId = parseInt(e.currentTarget.value);
		const selectedSport = test.data?.find(
			(sport) => sport.sportId === sportId
		);

		if (selectedSport) {
			dispatch(setSelectedSport(selectedSport));
		}

		// dispatch({
		// 	type: PLAYERS_ACTIONS.RESET_PLAYERS,
		// });

		// dispatch({
		// 	type: DROPDOWN_ACTIONS.RESET,
		// });

		// dispatch({
		// 	type: TABLE_ACTIONS.RESET,
		// });
	}

	return (
		<div className="relative">
			<Select
				options={(test.data ?? [])
					.filter(
						(sport) =>
							sport.isEnabled &&
							sport.hasPublicContests &&
							sport.supported
					)
					.sort((a, b) =>
						a.regionAbbreviatedSportName.localeCompare(
							b.regionAbbreviatedSportName
						)
					)
					.map(
						(sport) =>
							({
								value: sport.sportId,
								label: sport.regionAbbreviatedSportName,
							} as IValueLabel)
					)}
				value={sports.selectedSport?.sportId.toString() || ''}
				hideLabel
				id="selectSport"
				label="Select sport"
				placeholder="Select sport"
				disabled={!test.data}
				onChange={handleSportChange}
				testId="sports-select"
			/>
		</div>
	);
};

export default Sports;
