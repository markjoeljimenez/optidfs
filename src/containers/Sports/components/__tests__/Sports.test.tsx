import userEvent from '@testing-library/user-event';
import { logRoles, render, screen, waitFor } from '@/test/render';
import Sports from '../Sports';

const preloadedState = {
	providers: {
		provider: 'draftkings',
	},
};

describe('Sports', () => {
	it('should match snapshot', () => {
		// Arrange
		const sports = render(<Sports />);

		// Assert
		expect(sports).toMatchSnapshot();
	});

	it('should remain disabled if no provider is selected', async () => {
		// Arrange
		render(<Sports />);

		const sportsSelect = screen.getByTestId(
			'sports-select'
		) as HTMLSelectElement;

		// Assert
		expect(sportsSelect.hasAttribute('disabled')).toBe(true);
	});

	it('should render sports options if provider is selected', async () => {
		// Arrange
		render(<Sports />, {
			preloadedState,
		});

		const sportsSelect = screen.getByTestId(
			'sports-select'
		) as HTMLSelectElement;

		await waitFor(() =>
			expect(sportsSelect.hasAttribute('disabled')).toBe(false)
		);

		// Assert
		expect(sportsSelect.options[1].value).toBe('4');
		expect(sportsSelect.options[2].value).toBe('1');
	});

	it('should set selectedSport on change', async () => {
		// Arrange
		const { store } = render(<Sports />, {
			preloadedState,
		});

		const sportsSelect = screen.getByTestId(
			'sports-select'
		) as HTMLSelectElement;

		await waitFor(() =>
			expect(sportsSelect.hasAttribute('disabled')).toBe(false)
		);

		// Act
		userEvent.selectOptions(sportsSelect, ['4']);

		// Assert
		expect(store.getState().sports.selectedSport?.sportId).toBe(4);
	});
});
