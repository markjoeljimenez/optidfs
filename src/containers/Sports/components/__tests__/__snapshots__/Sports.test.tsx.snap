// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sports should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
      >
        <div>
          <div
            class=""
          >
            <label
              class="block text-sm font-medium text-gray-700 sr-only"
              for="selectSport"
            >
              Select sport
            </label>
          </div>
          <div
            class="flex"
          >
            <select
              class="focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded cursor-pointer"
              data-testid="sports-select"
              disabled=""
              id="selectSport"
            >
              <option
                disabled=""
                value=""
              >
                Select sport
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
    >
      <div>
        <div
          class=""
        >
          <label
            class="block text-sm font-medium text-gray-700 sr-only"
            for="selectSport"
          >
            Select sport
          </label>
        </div>
        <div
          class="flex"
        >
          <select
            class="focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded cursor-pointer"
            data-testid="sports-select"
            disabled=""
            id="selectSport"
          >
            <option
              disabled=""
              value=""
            >
              Select sport
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
